File: framework/src/play-java-ws/src/main/java/play/libs/ws/WSResponse.java
/*
 * Copyright (C) 2009-2016 Lightbend Inc. <https://www.lightbend.com>
 */
package play.libs.ws;

import com.fasterxml.jackson.databind.JsonNode;
import org.w3c.dom.Document;

import java.io.InputStream;
import java.net.URI;
import java.util.List;
import java.util.Map;

/**
 * This is the WS response from the server.
 */
public interface WSResponse {

    /**
     * @return all the headers from the response.
     */
    Map<String, List<String>> getAllHeaders();

    /**
     * @param key    the header's name
     * @return a single header value from the response.
     */
    String getHeader(String key);

    /**
     * @return the underlying implementation response object, if any.
     */
    Object getUnderlying();

    /**
     * @return the HTTP status code from the response.
     */
    int getStatus();

    /**
     * @return the text associated with the status code.
     */
    String getStatusText();

    /**
     * @return all the cookies from the response.
     */
    List<WSCookie> getCookies();

    /**
     * @param name    the cookie name
     * @return a single cookie from the response, if any.
     */
    WSCookie getCookie(String name);

    /**
     * @return the body as a string.
     */
    String getBody();

    /**
     * @return the body as XML.
     */
    Document asXml();

    /**
     * @return the body as JSON node.
     */
    JsonNode asJson();

    /**
     * @return the body as a stream.
     */
    InputStream getBodyAsStream();

    /**
     * @return the body as an array of bytes.
     */
    byte[] asByteArray();

    /**
     * @return the URI of the response.
     */
    URI getUri();
}

File: framework/src/play-ws/src/main/scala/play/api/libs/ws/WSResponse.scala
/*
 * Copyright (C) 2009-2018 Lightbend Inc. <https://www.lightbend.com>
 */

package play.api.libs.ws

import java.net.URI

import akka.stream.scaladsl.Source
import akka.util.ByteString
import play.api.libs.json.JsValue

import scala.xml.Elem

/**
 * A WS Response that can use Play specific classes.
 */
trait WSResponse extends StandaloneWSResponse with WSBodyReadables {

  /**
   * The response status code.
   */
  override def status: Int

  /**
   * The response status message.
   */
  override def statusText: String

  /**
   * Return the current headers for this response.
   */
  override def headers: Map[String, Seq[String]]

  /**
   * Get the underlying response object.
   */
  override def underlying[T]: T

  /**
   * Get all the cookies.
   */
  override def cookies: Seq[WSCookie]

  /**
   * Get only one cookie, using the cookie name.
   */
  override def cookie(name: String): Option[WSCookie]

  override def contentType: String = super.contentType

  override def header(name: String): Option[String] = super.header(name)

  override def headerValues(name: String): Seq[String] = super.headerValues(name)

  /**
   * The response body as the given type.  This renders as the given type.
   * You must have a BodyReadable in implicit scope, which is done with
   *
   * {{{
   * class MyClass extends play.api.libs.ws.WSBodyReadables {
   *   // JSON and XML body readables
   * }
   * }}}
   *
   * The simplest use case is
   *
   * {{{
   * val responseBodyAsString: String = response.getBody[String]
   * }}}
   *
   * But you can also render as JSON
   *
   * {{{
   * val responseBodyAsJson: JsValue = response.getBody[JsValue]
   * }}}
   *
   * or as XML:
   *
   * {{{
   * val xml: Elem = response.getBody[Elem]
   * }}}
   */
  override def body[T: BodyReadable]: T = super.body[T]

  /**
   * The response body as String.
   */
  override def body: String

  /**
   * The response body as a byte string.
   */
  override def bodyAsBytes: ByteString

  override def bodyAsSource: Source[ByteString, _]

  @deprecated("Use response.headers", "2.6.0")
  def allHeaders: Map[String, Seq[String]]

  def xml: Elem

  def json: JsValue

}
